libraries/bot-ai-qna/src/main/java/com/microsoft/bot/ai/qna/QnAMaker.java

```diff
@@ -0,0 +1,338 @@
+                                                                                             TurnContext turnContext,
+                                                                                             @Nullable Map<String, String> telemetryProperties,
+                                                                                             @Nullable Map<String, Double> telemetryMetrics) {
+        Map<String, String> properties = new HashMap<>();
+        Map<String, Double> metrics = new HashMap<>();
+
+        properties.put(QnATelemetryConstants.KNOWLEDGE_BASE_ID_PROPERTY, endpoint.getKnowledgeBaseId());
+
+        String text = turnContext.getActivity().getText();
+        String userName = turnContext.getActivity().getFrom().getName();
+
+        // Use the LogPersonalInformation flag to toggle logging PII data, text and user name are common examples
+        if (this.logPersonalInformation) {
+            if (!StringUtils.isWhitespace(text)) {
+                properties.put(QnATelemetryConstants.QUESTION_ID_PROPERTY, text);
+            }
+
+            if (!StringUtils.isWhitespace(userName)) {
+                properties.put(QnATelemetryConstants.USERNAME_PROPERTY, userName);
+            }
+        }
+
+        // Fill in QnA Results (found or not)
+        if (queryResults.length > 0) {
+            JacksonAdapter jacksonAdapter = new JacksonAdapter();
+            QueryResult queryResult = queryResults[0];
+            try {
+                properties.put(QnATelemetryConstants.MATCHED_QUESTION_PROPERTY,
+                    jacksonAdapter.serialize(queryResult.getQuestion()));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            properties.put(QnATelemetryConstants.QUESTION_ID_PROPERTY, queryResult.getId().toString());
+            properties.put(QnATelemetryConstants.ANSWER_PROPERTY, queryResult.getAnswer());
+            metrics.put(QnATelemetryConstants.SCORE_PROPERTY, queryResult.getScore().doubleValue());
+            properties.put(QnATelemetryConstants.ARTICLE_FOUND_PROPERTY, "true");
+        } else {
+            properties.put(QnATelemetryConstants.MATCHED_QUESTION_PROPERTY, "No Qna Question matched");
+            properties.put(QnATelemetryConstants.QUESTION_ID_PROPERTY, "No QnA Question Id matched");
+            properties.put(QnATelemetryConstants.ANSWER_PROPERTY, "No Qna Answer matched");
+            properties.put(QnATelemetryConstants.ARTICLE_FOUND_PROPERTY, "false");
+        }
+
+        // Additional Properties can override "stock" properties.
+        if (telemetryProperties != null) {
+            telemetryProperties.putAll(properties);
+
+        }
+
+        // Additional Metrics can override "stock" metrics.
+        if (telemetryMetrics != null) {
+            telemetryMetrics.putAll(metrics);
+        }
+
+        return CompletableFuture.completedFuture(new Pair(telemetryProperties, telemetryMetrics));
```

Comment:
This is not okay, you need to validate telemetryPrperties and telemetryMetrics.

See [here](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.AI.QnA/QnAMaker.cs#L342)